src/change_trust_asset.js

```diff
@@ -0,0 +1,280 @@
+ * @param {number} liquidityPoolParameters.fee – The liquidity pool fee. For now the only fee supported is `30`.
+ */
+export class ChangeTrustAsset {
+  constructor(code, issuer, liquidityPoolParameters) {
+    if (!code && !issuer && !liquidityPoolParameters) {
+      throw new Error(
+        'Must provide either code, issuer or liquidityPoolParameters'
+      );
+    }
+
+    if ((code || issuer) && liquidityPoolParameters) {
+      throw new Error(
+        'Must provide either code (and optionally issuer) or liquidityPoolParameters but not both'
+      );
+    }
+
+    // Validate code & issuer.
+    if ((code || issuer) && !/^[a-zA-Z0-9]{1,12}$/.test(code)) {
+      throw new Error(
+        'Asset code is invalid (maximum alphanumeric, 12 characters at max)'
+      );
+    }
+    if (code && String(code).toLowerCase() !== 'xlm' && !issuer) {
+      throw new Error('Issuer cannot be null');
+    }
+    if (issuer && !StrKey.isValidEd25519PublicKey(issuer)) {
+      throw new Error('Issuer is invalid');
+    }
+
+    this.code = code;
+    this.issuer = issuer;
+    if (code) {
+      // return early if it's an asset with code.
+      return;
+    }
+
+    // Validate liquidity pool params.
+    const { assetA, assetB, fee } = liquidityPoolParameters;
+    if (!assetA || !(assetA instanceof Asset)) {
+      throw new Error('assetA is invalid');
+    }
+    if (!assetB || !(assetB instanceof Asset)) {
+      throw new Error('assetB is invalid');
+    }
+    if (Asset.compare(assetA, assetB) !== -1) {
+      throw new Error('Assets are not in lexicographic order');
+    }
+    if (!fee || fee !== LiquidityPoolFeeV18) {
+      throw new Error('fee is invalid');
+    }
+    this.liquidityPoolParameters = liquidityPoolParameters;
+  }
+
+  /**
+   * Returns a change trust asset object of the type "credit asset". Can be a
+   * native or an issued asset.
+   * @static
+   * @param {string} code - The asset code.
+   * @param {string} issuer - The account ID of the asset issuer.
+   * @return {ChangeTrustAsset}
+   * @memberof ChangeTrustAsset
+   */
+  static creditAsset(code, issuer) {
+    return new ChangeTrustAsset(code, issuer);
+  }
+
+  /**
+   * Returns a change trust asset object of the type "liquidity pool shares".
+   * @static
+   * @param {LiquidityPoolParameters} liquidityPoolParameters – The liquidity pool parameters.
+   * @param {Asset} liquidityPoolParameters.assetA – The first asset in the Pool, it must respect the rule assetA < assetB.
+   * @param {Asset} liquidityPoolParameters.assetB – The second asset in the Pool, it must respect the rule assetA < assetB.
+   * @param {number} liquidityPoolParameters.fee – The liquidity pool fee. For now the only fee supported is `30`.
+   * @return {ChangeTrustAsset}
+   * @memberof ChangeTrustAsset
+   */
+  static liquidityPoolShare(liquidityPoolParameters) {
+    return new ChangeTrustAsset(undefined, undefined, liquidityPoolParameters);
```

Comment:
Okay, I'm not sure why I didn't consider this before, but can we step back for a second: **why are we even trying to mimic the XDR like this?** Just because the XDR makes this bizarre design choice in which it can hold the native asset, 4-char assets, 12-char assets, **or** liquidity pool parameters doesn't mean we need to follow that. 

We should provide user-friendly APIs, instead, so if `Asset(code, issuer)` is *exactly the same* as `ChangeTrustAsset(code, issuer)` (right?), why not *only* have something like `LiquidityPoolAsset(assetA, assetB, fee)` and just make people use `Asset` when they want, well, an asset (either native or code:issuer)? 

And if binary compatibility is a concern, we can have something like an `Asset.asChangeTrustXDR()` helper, instead.

Reply:
Wow, that's a great idea 👏 ! I'll play around with that and ask for your review later today.

Also, thanks a lot for putting some extra thought into this, it will make the implementation and usage much cleaner and remove the breaking changes! 🎉